pwndbg/heap_tracking.py

```diff
@@ -0,0 +1,357 @@
+operation, this module will only do so when it determines its view of the chunks
+has diverged from the one in libc in a way that would affect behavior. When such
+a diffence is detected, this module will rebuild the chunk maps in the range it
+determines to have been affected.
+
+Currently, the way it does this is by deleting and querying from libc the new
+status of all chunks that overlap the region of a new allocation when it detects
+that allocation overlaps chunks it previously considered free.
+
+This approach lets us avoid a lot of the following linked lists that comes with
+trying to answer the allocation status of a chunk, by keeping at hand as much
+known-good information as possible about them. Keep in mind that, although it is
+much faster than going to libc every time we need to know the allocation status
+of a chunk, this approach does have drawbacks when it comes to memory usage.
+
+# Compatibility
+Currently module assumes the inferior is using GLibc.
+
+There are points along the code in this module where the assumptions it makes
+are explicitly documented and checked to be valid for the current inferior, so
+that it may be immediately clear to the user that something has gone wrong if
+they happen to not be valid. However, be aware that there may be assumptions
+that were not made explicit.
+
+"""
+
+import gdb
+from sortedcontainers import SortedDict
+
+import pwndbg.gdblib
+
+MALLOC_NAME = "malloc"
+FREE_NAME = "free"
+
+
+class FreeChunkWatchpoint(gdb.Breakpoint):
+    def __init__(self, chunk, tracker):
+        self.chunk = chunk
+        self.tracker = tracker
+        super().__init__(
+            f"*(char[{chunk.size}]*){chunk.address:#x}", type=gdb.BP_WATCHPOINT, internal=True
+        )
+
+    def stop(self):
+        pwndbg.gdblib.regs.__getattr__.cache.clear()
+        if not in_program_code_stack():
+            # Untracked.
+            return False
+
+        # malloc() and free() implementations will often modify the data in the
+        # payload of a freed chunk, where the watchpoint is insalled. So, we
+        # should not flag accesses done as a result of a call to either.
+        if self.tracker.is_performing_memory_management():
+            # We explicitly allow this operation.
+            return False
+
+        return True
+
+
+class AllocChunkWatchpoint(gdb.Breakpoint):
+    def __init__(self, chunk):
+        pwndbg.gdblib.regs.__getattr__.cache.clear()
+
+        self.chunk = chunk
+        super().__init__(f"*(char[{chunk.size}]*){chunk.address:#x}", internal=True)
+
+    def stop(self):
+        return False
+
+
+class Chunk:
+    def __init__(self, address, size, requested_size, flags):
+        self.address = address
+        self.size = size
+        self.requested_size = requested_size
+        self.flags = flags
+
+
+class Tracker:
+    def __init__(self):
+        self.free_chunks = SortedDict()
+        self.alloc_chunks = SortedDict()
+        self.free_wps = dict()
```

Comment:
Nitpick: I had to scroll a fair way to find what `free_wps` stood for, ain't nothing wrong with descriptive variable names.

Reply:
Fair enough, that's been changed to `free_watchpoints` now.